pkg/signatures/cosign_sig_verifier.go

```diff
@@ -68,54 +80,175 @@ func newCosignSignatureVerifier(config *storage.CosignPublicKeyVerification) (*c
+		// let's first try to go through any options that might have been successfully created (i.e. the key ones)
+		// and attempt to verify the signature.
+		allVerifyErrs = multierror.Append(allVerifyErrs, err)
+	}
+
+	for _, opts := range c.verifierOpts {
+		verifiedImageReferences, err := verifyImageSignatures(ctx, sigs, hash, image, opts)
+		if err == nil && len(verifiedImageReferences) != 0 {
+			return storage.ImageSignatureVerificationResult_VERIFIED, verifiedImageReferences, nil
+		}
+		allVerifyErrs = multierror.Append(allVerifyErrs, err)
+	}
+
+	return storage.ImageSignatureVerificationResult_FAILED_VERIFICATION, nil, allVerifyErrs
+}
+
+func (c *cosignSignatureVerifier) createVerifierOpts() error {
+	var verifierErrs error
+
 	for _, key := range c.parsedPublicKeys {
 		// For now, only supporting SHA256 as algorithm.
 		v, err := signature.LoadVerifier(key, crypto.SHA256)
 		if err != nil {
-			allVerifyErrs = multierror.Append(allVerifyErrs, errors.Wrap(err, "creating verifier"))
+			verifierErrs = multierror.Append(verifierErrs, errors.Wrap(err, "creating verifier"))
 			continue
 		}
+		opts := defaultCosignCheckOpts()
 		opts.SigVerifier = v
-		verifiedImageReferences, err := verifyImageSignatures(ctx, sigs, hash, image, opts)
-		if err == nil && len(verifiedImageReferences) != 0 {
-			return storage.ImageSignatureVerificationResult_VERIFIED, verifiedImageReferences, nil
+		c.verifierOpts = append(c.verifierOpts, opts)
+	}
+
+	for _, certs := range c.certs {
+		opts, err := cosignCheckOptsFromCert(certs)
+		if err != nil {
+			verifierErrs = multierror.Append(verifierErrs, errors.Wrap(err, "creating cosign check opts"))
+			continue
 		}
-		allVerifyErrs = multierror.Append(allVerifyErrs, err)
+		c.verifierOpts = append(c.verifierOpts, opts)
 	}
 
-	return storage.ImageSignatureVerificationResult_FAILED_VERIFICATION, nil, allVerifyErrs
+	return verifierErrs
+}
+
+func defaultCosignCheckOpts() cosign.CheckOpts {
+	return cosign.CheckOpts{
+		ClaimVerifier: cosign.SimpleClaimVerifier,
+		// With the latest version of cosign, by default signatures will be uploaded to rekor.
+		// This means that also during verification, an entry in the transparency log will be expected and verified.
+		// Since currently this is not the case in what we support / offer, explicitly disable this for now
+		// until we enable and expect this to be the case.
+		IgnoreSCT:  true,
+		IgnoreTlog: true,
+	}
+}
+
+func cosignCheckOptsFromCert(cert certVerificationData) (cosign.CheckOpts, error) {
+	opts := defaultCosignCheckOpts()
+
+	// Skip verifying the identities when the wildcard matching logic being used. This fixes an issue
+	// with verifying the identity which will yield an error when using the wildcard expressions _and_ the certificate
+	// to verify has no identities associated with it (i.e. within the BYOPKI use-case).
+	if cert.oidcIssuerExpr != ".*" && cert.identityExpr != ".*" {
```

Comment:
Is it enough to check for wildcards this way? What about more complex regex including wildcards, like `id-.*`?

Reply:
I made the comment more clear, it's basically when it's matching "everything". This is suggested when using to verify a certificate which e.g. has no identity set. Right now, cosign will fail in for the identity part, not for the issuer.

So, that's why this is here, other wildcard matching will be handled as-is, and this is only to allow verifying signatures where the certificate has no identity being set.

I've also contemplated just dropping this completely and say "hey you're not using a strictly compliant fulcio certificate where all SAN extension are set", which would be an alternative to having this here, but I felt from a UX perspective it would be better to allow verifying certs where you don't necessarily have to set the SAN extensions.